src/system/SystemObject.h

```diff
@@ -241,185 +347,130 @@ class ObjectPool
+template <typename T>
+class ObjectPoolDynamic : public ObjectPoolBase<T>
 {
-    T * lReturn = nullptr;
+public:
+    T * TryCreate() override
+    {
 
-    (void) static_cast<Object *>(lReturn); /* In C++-11, this would be a static_assert that T inherits Object. */
+        T * newNode = new T();
 
-#if CHIP_SYSTEM_CONFIG_POOL_USE_HEAP
-    T * newNode = new T();
+        if (newNode->TryCreate(sizeof(T)))
+        {
+            std::lock_guard<std::mutex> lock(mMutex);
+            Object * p = &mDummyHead;
+            if (p->mNext)
+            {
+                p->mNext->mPrev = newNode;
+            }
+            newNode->mNext     = p->mNext;
+            p->mNext           = newNode;
+            newNode->mPrev     = p;
+            newNode->mMutexRef = &mMutex;
+            return newNode;
+        }
+        else
+        {
+            delete newNode;
+            return nullptr;
+        }
+    }
 
-    if (newNode->TryCreate(sizeof(T)))
+    /**
+     * @brief
+     *   Run a functor for each active object in the pool
+     *
+     *  @param     function The functor of type `bool (*)(T*)`, return false to break the iteration
+     *  @return    bool     Returns false if broke during iteration
+     */
+    template <typename Function>
+    bool ForEachActiveObject(Function && function)
     {
         std::lock_guard<std::mutex> lock(mMutex);
-        Object * p = &mDummyHead;
-        if (p->mNext)
+        Object * p = mDummyHead.mNext;
+        while (p)
         {
-            p->mNext->mPrev = newNode;
+            if (!function(static_cast<T *>(p)))
+            {
+                return false;
+            }
+            p = p->mNext;
         }
-        newNode->mNext     = p->mNext;
-        p->mNext           = newNode;
-        newNode->mPrev     = p;
-        newNode->mMutexRef = &mMutex;
-        lReturn            = newNode;
+        return true;
     }
-    else
+    void Reset() override
     {
-        delete newNode;
-    }
-#else // CHIP_SYSTEM_CONFIG_POOL_USE_HEAP
-    unsigned int lIndex = 0;
-
-    for (lIndex = 0; lIndex < N; ++lIndex)
-    {
-        T & lObject = reinterpret_cast<T *>(mArena.uMemory)[lIndex];
+        std::lock_guard<std::mutex> lock(mMutex);
+        Object * p = mDummyHead.mNext;
 
-        if (lObject.TryCreate(sizeof(T)))
+        while (p)
         {
-            lReturn = &lObject;
-            break;
+            Object * del = p;
+            p            = p->mNext;
+            delete del;
         }
-    }
 
-#if CHIP_SYSTEM_CONFIG_PROVIDE_STATISTICS
-    unsigned int lNumInUse = 0;
-
-    if (lReturn != nullptr)
-    {
-        lIndex++;
-        lNumInUse = lIndex;
-        GetNumObjectsInUse(lIndex, lNumInUse);
+        mDummyHead.mNext = nullptr;
     }
-    else
-    {
-        lNumInUse = N;
-    }
-
-    UpdateHighWatermark(lNumInUse);
-#endif
+    void GetStatistics(chip::System::Stats::count_t & aNumInUse, chip::System::Stats::count_t & aHighWatermark) override {}
```

Comment:
I know this is preserving existing behavior, but that behavior is broken: it silently leaves outparams un-initialized.  Followup for that, please?

Reply:
https://github.com/project-chip/connectedhomeip/issues/11490